homeassistant/components/connectedcars/sensor.py

```diff
@@ -0,0 +1,167 @@
+"""Support for reading vehicle status from ConnectedCars.io."""
+from typing import Any, Callable, Dict, List
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ATTRIBUTION,
+    ATTR_DEVICE_CLASS,
+    ATTR_NAME,
+    LENGTH_KILOMETERS,
+    PERCENTAGE,
+    VOLT,
+    VOLUME_LITERS,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+
+from . import ConnectedCarsDataUpdateCoordinator
+from .const import (
+    ATTR_API_VEHICLE_FUELLEVEL,
+    ATTR_API_VEHICLE_FUELPERCENTAGE,
+    ATTR_API_VEHICLE_LICENSEPLATE,
+    ATTR_API_VEHICLE_MAKE,
+    ATTR_API_VEHICLE_MODEL,
+    ATTR_API_VEHICLE_NAME,
+    ATTR_API_VEHICLE_ODOMETER,
+    ATTR_API_VEHICLE_VIN,
+    ATTR_API_VEHICLE_VOLTAGE,
+    ATTR_ICON,
+    ATTR_IDENTIFIERS,
+    ATTR_LABEL,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    ATTR_UNIT,
+    ATTRIBUTION,
+    DOMAIN,
+)
+
+SENSOR_TYPES = {
+    ATTR_API_VEHICLE_ODOMETER: {
+        ATTR_DEVICE_CLASS: None,
+        ATTR_ICON: "mdi:trending-up",
+        ATTR_LABEL: ATTR_API_VEHICLE_ODOMETER,
+        ATTR_UNIT: LENGTH_KILOMETERS,
+    },
+    ATTR_API_VEHICLE_FUELLEVEL: {
+        ATTR_DEVICE_CLASS: None,
+        ATTR_ICON: "mdi:gauge",
+        ATTR_LABEL: ATTR_API_VEHICLE_FUELLEVEL,
+        ATTR_UNIT: VOLUME_LITERS,
+    },
+    ATTR_API_VEHICLE_FUELPERCENTAGE: {
+        ATTR_DEVICE_CLASS: None,
+        ATTR_ICON: "mdi:gas-station",
+        ATTR_LABEL: ATTR_API_VEHICLE_FUELPERCENTAGE,
+        ATTR_UNIT: PERCENTAGE,
+    },
+    ATTR_API_VEHICLE_VOLTAGE: {
+        ATTR_DEVICE_CLASS: None,
+        ATTR_ICON: "mdi:flash",
+        ATTR_LABEL: ATTR_API_VEHICLE_VOLTAGE,
+        ATTR_UNIT: VOLT,
+    },
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    config_entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity], bool], None],
+) -> None:
+    """Set up Conncted Cars entities based on a config entry."""
+    name = config_entry.title
+
+    coordinator: ConnectedCarsDataUpdateCoordinator = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+
+    # Maybe add vin to unique id
+    sensors = []
+    for sensor in SENSOR_TYPES:
+        unique_id = f"{config_entry.unique_id}-{sensor.lower()}"
+        sensors.append(ConnectedCarsSensor(coordinator, name, sensor, unique_id))
+
+    async_add_entities(sensors, False)
+
+
+class ConnectedCarsSensor(Entity):
```

Comment:
Extend from `CoordinatorEntity` is has a lot of pre-defined logic already, allowing you to clean up this class quite a bit.